src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -225,48 +312,135 @@ public static unsafe OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Spa
         {
-            if (buffer.IsEmpty)
+            Unsafe.SkipInit(out bytesConsumed);
+            Unsafe.SkipInit(out bytesWritten);
+
+            const int BlockSize = 4;
+            Span<byte> buffer = stackalloc byte[BlockSize];
+            OperationStatus status = OperationStatus.Done;
+
+            while (!utf8.IsEmpty)
             {
-                bytesWritten = 0;
-                return OperationStatus.Done;
+                int encodedIdx = 0;
+                int bufferIdx = 0;
+                int skipped = 0;
+
+                for (; encodedIdx < utf8.Length && (uint)bufferIdx < (uint)buffer.Length; ++encodedIdx)
+                {
+                    if (IsWhitespace(utf8[encodedIdx]))
+                    {
+                        skipped++;
+                    }
+                    else
+                    {
+                        buffer[bufferIdx] = utf8[encodedIdx];
+                        bufferIdx++;
+                    }
+                }
+
+                utf8 = utf8.Slice(encodedIdx);
+                bytesConsumed += skipped;
+
+                if (bufferIdx == 0)
+                {
+                    continue;
+                }
+
+                bool hasAnotherBlock = utf8.Length >= BlockSize && bufferIdx == BlockSize;
+                bool localIsFinalBlock = !hasAnotherBlock;
+
+                // If this block contains padding and there's another block, then only whitespace may follow for being valid.
+                if (hasAnotherBlock)
+                {
+                    int paddingCount = GetPaddingCount(ref buffer[^1]);
+                    if (paddingCount > 0)
+                    {
+                        hasAnotherBlock = false;
+                        localIsFinalBlock = true;
+                    }
+                }
+
+                if (localIsFinalBlock && !isFinalBlock)
+                {
+                    localIsFinalBlock = false;
+                }
+
+                status = DecodeFromUtf8Core(buffer.Slice(0, bufferIdx), bytes, out int localConsumed, out int localWritten, localIsFinalBlock);
+                bytesConsumed += localConsumed;
+                bytesWritten += localWritten;
+
+                if (status != OperationStatus.Done)
+                {
+                    return status;
+                }
+
+                // The remaining data must all be whitespace in order to be valid.
+                if (!hasAnotherBlock)
+                {
+                    for (int i = 0; i < utf8.Length; ++i)
+                    {
+                        if (!IsWhitespace(utf8[i]))
+                        {
+                            // Revert previous dest increment, since an invalid state followed.
+                            bytesConsumed -= localConsumed;
+                            bytesWritten -= localWritten;
```

Comment:
IMO this is not correct, as whitespace should be treated as consumed, see https://github.com/gfoidl/Base64/pull/184#issuecomment-1363806704

Reply:
Whitespace is being counted, this is when it is not whitespace:   
(https://github.com/dotnet/runtime/pull/79334/files/fb9e8de707c69361e286bf11c401e8d1c7718a86#diff-db463201901c2d83d2b563871ae11fafee9d5afe94e4d014b77212996b25f770R402)

```C#
if (!IsWhitespace(utf8[i]))
{
    // Revert previous dest increment, since an invalid state followed.
    bytesConsumed -= localConsumed;
    bytesWritten -= localWritten;

    return OperationStatus.InvalidData;
}
```

This was to allow existing tests to pass to maintain existing behavior. If the decrements are removed, DecodingInvalidBytesPadding fails.   https://github.com/dotnet/runtime/blob/8da03fa38a8e690af7443702e42042091412a130/src/libraries/System.Memory/tests/Base64/Base64DecoderUnitTests.cs#L413